WcaOnRails/app/models/person.rb

```diff
@@ -1,13 +1,93 @@
+  has_many :ranksSingle, primary_key: "wca_id", foreign_key: "personId", class_name: "RanksSingle"
+
+  scope :current, -> { where(subId: 1) }
+
+  validates :name, presence: true
+  validates :countryId, presence: true
+
+  before_validation :unpack_dob
+  private def unpack_dob
+    if @dob.nil? && !dob.blank?
+      @dob = dob.strftime("%F")
+    end
+    if @dob.blank?
+      self.year = self.month = self.day = 0
+    else
+      unless @dob =~ /\A\d{4}-\d{2}-\d{2}\z/
+        errors.add(:dob, "is invalid")
+        return false
+      end
+      self.year, self.month, self.day = @dob.split("-").map(&:to_i)
+      unless Date.valid_date? self.year, self.month, self.day
+        errors.add(:dob, "is invalid")
+        return false
+      end
+    end
+  end
 
-  alias_method :wca_id, :id
+  validate :dob_must_be_in_the_past
+  private def dob_must_be_in_the_past
+    if dob && dob >= Date.today
+      errors.add(:dob, "must be in the past")
+    end
+  end
+
+  # If someone represented country A, and now represents country B, it's
+  # easy to tell which solves are which (results have a countryId).
+  # Fixing their country (B) to a new country C is easy to undo, just change
+  # all Cs to Bs. However, if someone accidentally fixes their country from B
+  # to A, then we cannot go back, as all their results are now for country A.
+  validate :cannot_change_country_to_country_represented_before, if: :countryId_changed?, unless: -> { new_record? || @updating_using_sub_id }
+  private def cannot_change_country_to_country_represented_before
+    has_represented_this_country_already = Person.exists?(wca_id: wca_id, countryId: countryId)
+    if has_represented_this_country_already
+      errors.add(:countryId, "Cannot change the country to a country the person have already represented in the past.")
+    end
+  end
+
+  # This is necessary because we use a view instead of a real table.
+  # Using `select` statement with `id` column causes MySQL to set a default value of 0,
+  # so creating a Person returns the new record with id = 0, making the record reference 'died'.
+  # The workaround is to set id attribute to nil before the object is created and let Rails reload it after creation.
+  # For reference: https://github.com/rails/rails/issues/5982
+  before_create -> { self.id = nil }
+
+  after_update :update_results_table_and_associated_user
+  private def update_results_table_and_associated_user
+    unless @updating_using_sub_id
+      results.where(personName: name_was).update_all(personName: name) if name_changed?
+      results.where(countryId: countryId_was).update_all(countryId: countryId) if countryId_changed?
+    end
+    user.save! if user # User copies data from the person before validation, so this will update him.
+  end
+
+  # Keep the information since admin_controller needs it.
+  attr_reader :country_id_changed
+  after_update -> { @country_id_changed = countryId_changed? }
```

Comment:
Just wanted to keep this out of controller, not sure what's more appropriate here.